ssg/templates.py

```diff
@@ -236,42 +208,32 @@ def process_product_vars(self, all_variables):
 
         return processed
 
-    def build_rule(self, rule_id, rule_title, template, langs_to_generate, identifiers,
-                   platforms=None):
+    def render_lang_file(self, template_name, template_vars, lang, local_env_yaml):
         """
-        Builds templated content for a given rule for selected languages,
-        writing the output to the correct build directories.
+        Builds and returns templated content for a given rule for a given
+        language; does not write the output to disk.
         """
-        template_name = self.get_template_name(template, rule_id)
-        try:
-            template_vars = self.process_product_vars(template["vars"])
-        except KeyError:
-            raise ValueError(
-                "Rule {0} does not contain mandatory 'vars:' key under "
-                "'template:' key.".format(rule_id))
-        # Add the rule ID which will be reused in OVAL templates as OVAL
-        # definition ID so that the build system matches the generated
-        # check with the rule.
-        template_vars["_rule_id"] = rule_id
-        # checks and remediations are processed with a custom YAML dict
-        local_env_yaml = self.env_yaml.copy()
-        local_env_yaml["rule_id"] = rule_id
-        local_env_yaml["rule_title"] = rule_title
+        if lang not in self.templates[template_name].langs:
+            return None
+
+        template_file_name = lang.name + ".template"
+        template_file_path = os.path.join(self.templates_dir, template_name, template_file_name)
+        template_parameters = self.templates[template_name].preprocess(template_vars, lang.name)
         local_env_yaml["products"] = self.env_yaml["product"]
-        if identifiers is not None:
-            local_env_yaml["cce_identifiers"] = identifiers
+        env_yaml = self.env_yaml.copy()
+        env_yaml.update(local_env_yaml)
+        jinja_dict = ssg.utils.merge_dicts(env_yaml, template_parameters)
+        try:
+            filled_template = ssg.jinja.process_file_with_macros(template_file_path, jinja_dict)
+        except Exception as e:
+            print("Error in template: %s (lang: %s)" % (template_name, lang.name))
+            raise e
 
-        for lang in langs_to_generate:
-            try:
-                self.build_lang(
-                    rule_id, template_name, template_vars, lang, local_env_yaml, platforms)
-            except Exception as e:
-                raise e(
-                    "Error building templated {0} content for rule {1}".format(lang, rule_id))
+        return filled_template
 
-    def get_lang_for_rule(self, rule_id, rule_title, template, language):
+    def get_lang_contents(self, rule_id, rule_title, template, language, extra_env=None):
```

Comment:
Let's get rid of the name `extra_env`, because on this level of abstraction, we don't care about origins of the dictionary. Isn't it just a dictionary with keys and substitutions?

Reply:
That's a drawback of merging a feature step-by-step. You don't see full picture. It will be used for platforms in a different way later.